Doc for persisted queries (#496)
apollographql/docs: src/content/graphos/basics/operations/persisted-queries.mdx

---
title: Operation safelisting with persisted queries
description: Reject unrecognized operations while minimizing request latency
---

<blockquote>

**This feature is available only with a [GraphOS Enterprise plan](/graphos/enterprise/).** It is currently in [preview](/resources/product-launch-stages#preview).

If your organization _doesn't_ currently have an Enterprise plan, you can test out this functionality by signing up for a free [Enterprise trial](/graphos/org/plans/#enterprise-trials).

</blockquote>

With [GraphOS Enterprise](../enterprise/), you can enhance your graph's security by maintaining a **persisted query list** (or **PQL**) for your supergraph's self-hosted router. This list should contain _all_ of the trusted GraphQL operations that _all_ of your app clients execute against your router:

```mermaid
graph LR;

  subgraph "CI/CD"
  webClient("Web client<br/>build");

>   androidClient("Android client<br/>build");
>   end

Reviewer: ```suggestion
  androidClient("Android client<br/>build");
  iosClient("iOS client<br/>build");
  end
```

  subgraph "GraphOS"
  pql[[Persisted<br/>Query List]];
  end
  
  subgraph "Your backend"
  router(["Apollo<br/>Router"]);
  end

  webClient & androidClient -->|Publishes<br/>operations|pql

>   router -.->|Fetches PQL<br/>at runtime|pql

Reviewer 2: This was the best I could get the Mermaid diagram looking, though it wasn't exactly what I was going for:

I'd love to have all clients point to the router node and that node point to all graphQL APIs. 

Is the current version good enough?

##### Apollo Client Web

1. In your app's project, install the `@apollo/generate-persisted-query-manifest` package as a dev dependency:

    ```bash
    npm install --save-dev @apollo/generate-persisted-query-manifest
    ```

    This package includes a CLI command you can run to generate a manifest file from your application source.

2. Generate your first manifest with the following command:

    ```bash
    npx generate-persisted-query-manifest
    ```

    - **If the command succeeds,** your manifest is written to `persisted-query-manifest.json`.
    - **If the command fails** (or if your manifest doesn't include all of the operations you expect it to), you can configure the command's behavior using the options described in the [package README](https://www.npmjs.com/package/@apollo/generate-persisted-query-manifest).

##### Apollo Kotlin

> Manifest generation requires Apollo Kotlin `3.8.2` or later.

To generate an operation manifest with Apollo Kotlin, you modify your project's [Gradle plugin configuration](/kotlin/advanced/plugin-configuration) to generate a manifest in addition to standard Kotlin source for your operations:

```kotlin
apollo {

>   service("myapi") {
>     operationManifestFormat.set("persistedQueryManifest") //highlight-line

Reviewer: `packageName` is mandatory in the configuration so I'd add it to avoid showing something that doesn't build

```suggestion
  service("myapi") {
    packageName.set("com.example.myapi")
    operationManifestFormat.set("persistedQueryManifest") //highlight-line
```

  }
}
```

> 

Reviewer: ```suggestion
The manifest will be generated in `build/generated/manifest/apollo/myapi/persistedQueryManifest.json`
```

1. It publishes all operations in the provided manifest file to whichever PQL is linked to the specified variant.
    - Publishing a manifest to a PQL is additive. Any _existing_ entries in the PQL remain.
    - If you publish an operation with the same `id` but different details from an existing entry in the PQL, the entire publish command fails with an error.

2. It updates any _other_ variants that the PQL is applied to, so routers associated with those variants can fetch their updated PQL.

### 5. Configure your router

After you [publish all operations](#4-publish-manifests-to-the-pql) to your PQL, you can configure your router instances to fetch and _use_ your PQL.

1. As mentioned in [Prerequisites](#0-prerequisites):
    - Make sure you're using version `1.25.0` or later of the Apollo Router.
    - Make sure your router instances are [connected to your GraphOS Enterprise organization](/router/enterprise-features/#enabling-enterprise-features), and that they're associated with a variant that your PQL is linked to.

2. Set your desired persisted query behavior in your router's YAML config file. For supported options, see [Router security levels](#router-security-levels).

> 3. Deploy your updated router instances to begin using your PQL.

Reviewer: I was expecting a section about "local development". What should I do during local development when I'm iterating on queries? Is there a setting for that? Should I publish a separate query list? Or (more likely), should I use a staging instance of my API that allows queries?

To use persisted queries _only_ to reduce network bandwidth and latency (_not_ for safelisting), add the following minimal configuration:

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
apq:
  enabled: false # APQ must be disabled 
```

This enables your clients to execute operations by providing their PQL-specified ID (instead of the full query string). Your router continues to accept full query strings as well, even for operations that _don't_ appear in its PQL.

### Safelisting (query strings allowed)

> ⚠️ **Before you apply this configuration,** first make sure your PQL contains _all_ of the GraphQL operations that _all active versions_ of your clients execute. If you enable safelisting _without_ ensuring this, your router will reject your unpublished client operations!

With the following configuration, your router allows _only_ GraphQL operations that are present in its PQL, while rejecting all other operations:

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
  safelist:
    enabled: true

>     freeform_graphql: 

Reviewer 3: The persisted query config looks like this:

```
preview_persisted_queries:
  enabled: true
  log_unknown: true
  safelist:
    require_id: true
    enabled: true
```

`freeform_graphql` needs to change to `require_id`. In addition they have the opposite meanings so other text may need updating.

Reviewer 2: Thanks @BrynCooke .

To confirm, this config allows safelisting with query strings:

```
preview_persisted_queries:
  enabled: true
  log_unknown: true
  safelist:
    require_id: false
    enabled: true
 ```
 
 and this config allows safelisting with IDs only:
 ```
 preview_persisted_queries:
  enabled: true
  log_unknown: true
  safelist:
    enabled: true
    require_id: true
``` 
 
 LMK if I've misunderstood.

      enabled: true # highlight-line
apq:
  enabled: false # APQ must be disabled 
```

> To execute an operation, clients can provide either its full operation string _or_ its PQL-specified ID.

Reviewer: Might be worth adding something saying this is working because the id is the sha256 hash and the server can get it from the query string if needed?

Reviewer 2: By "server" I'm assuming we're talking about the router deriving the hash from the query string if needed?

**This string must match _exactly_ the query document that your client sends your router, including whitespace and field order.** For details, see [Ensuring consistent operation signatures](#ensuring-consistent-operation-signatures).

</td>
</tr>

<tr>
<td>

###### `name`

</td>
<td>

The operation's name. Must match the name specified in `body`.

This value does _not_ need to be unique among operations in the PQL. Often, different clients execute slightly different operations with the same name, and those operations each require a separate entry in the PQL.

</td>
</tr>

<tr>
<td>

###### `type`

</td>
<td>

The type of GraphQL operation. Always one of the following values:

- `query`
- `mutation`
- `subscription`

</td>
</tr>

</tbody>
</table>

### Generating IDs

When generating operation IDs for a manifest, you should use a value that's unique to each operation, such as the query document's cryptographic hash. Apollo's [manifest generation tools](#3-generate-operation-manifests) use the base16 representation of the document's SHA256 hash, which is the same format used for [APQ](#differences-from-apq).

> By generating identifiers based on query documents this way, you ensure that different operations always have different IDs. This prevents unexpected collisions in your PQL.

Reviewer: Maybe loop back saying this is what allows the `Safelisting (query strings allowed)` mode?

```mermaid
flowchart LR
  subgraph Apps["First-party apps"]
  direction LR
  webClient("Web client");
  androidClient("Android client");
  iosClient("iOS client");
  end

    subgraph Router["Apollo Router"]
        direction TB
        router(["Graph<br/>Router"]);
        pql[[Persisted<br/>Query List]];
        router <--> pql
    end

  subgraph Supergraph
  direction LR
  api1(["GraphQL API"]);
  api2(["GraphQL API"]);
  api3(["GraphQL API"]);

  end
  Apps -->|Queries| Router -->|Trusted<br/> queries| Supergraph
```

Your router can use its persisted query list (PQL) to both **protect your supergraph** and **speed up your clients' operations**:

- When you enable **safelisting**, your router _rejects_ any incoming operations not registered in its PQL.
- Client apps can execute an operation by providing its PQL-specified ID _instead of_ an entire operation string.
    - Querying by ID can significantly reduce latency and bandwidth usage for very large operation strings.
    - Your router can _require_ that clients provide operations by ID and _reject_ full operation strings (even operation strings present in the PQL).

## Differences from automatic persisted queries

> The Apollo Router supports a related feature called [automatic persisted queries](/router/configuration/in-memory-caching#caching-automatic-persisted-queries-apq) (APQ). With APQ, clients can execute a GraphQL operation by sending the SHA256 hash of its operation string instead of the entire string.

Reviewer 2: I tend to avoid using "also" and consider it a weasel word, but it does adding something here.

> The Apollo Router supports a related feature called [automatic persisted queries](/router/configuration/in-memory-caching#caching-automatic-persisted-queries-apq) (APQ). With APQ, clients can execute a GraphQL operation by sending the SHA256 hash of its operation string instead of the entire string.

Reviewer 4: Sure, I bet you can come up with better ways to word this; my point (which I think you see) was just that the previous wording may make you believe that the Router *only* supports APQs and that you need something else for this feature.

      </td>
      <td>App developers</td>
    </tr>
    <tr>
      <td><a href="#4-client-updates">4. Client updates (Optional) </a> </td>
      <td>
        Update clients to use operation IDs rather than full query strings.
        <br/>
        <br/>
        <em>This step provides performance benefits but isn't necessary for safelisting.</em>
      </td>
      <td>App developers</td>
    </tr>
  </tbody>
</table>

Continue on for each step's details, or skip to the [incremental adoption section](#incremental-adoption-path) for the recommended for incremental adoption strategy. (This section assumes you have a high-level understanding of each implementation step.)

### 1. PQL creation and linking

To use persisted queries, you first need a **persisted query list** (**PQL**) in GraphOS Studio.
Platform teams create an empty PQL in GraphOS Studio so that client teams can register trusted operations to it.

Each PQL is associated with a single graph in GraphOS.
You can link a PQL to any number of different variants of its graph.
Although many variants may use the same PQL, each variant can only have one linked PQL at a time.

#### 1.1 PQL creation

> 1. From your organization's Graphs page in [GraphOS Studio](https://studio.apollographql.com/?referrer=docs-content), open the PQL page for a graph by clicking its PQL button:

Reviewer 4: (not a preview blocker imo)

> 1. From your organization's Graphs page in [GraphOS Studio](https://studio.apollographql.com/?referrer=docs-content), open the PQL page for a graph by clicking its PQL button:

Reviewer 2: Due to my inexperience with Studio, not sure which is the easier UX for most users.
For users with relatively few graphs (don't know how many graphs most users have), scrolling might not be an issue?

I can't find a way to dynamically link a graph setting's page. (E.g., if were I go to `https://studio.apollographql.com/graph/settings/graph/general` , it would automatically redirect to `https://studio.apollographql.com/graph/mschreiber-vyq62/variant/main/settings/graph/general`, the URL of my most recently opened graph's settings page.) Is that something we can do?

If not, I'd prefer to just keep the instructions as is (with a callout that you can also access PQLs from a graph's settings page), to keep the instructions more concise. (Otherwise, it seems there would be additional steps? "From GraphOS studio, go to the graph you want to open PQLs for, open any variant, go to "Settings", go to Graph settings rather than the variant's settings, etc.." LMK if there's a more concise way to direct users there...)

> 1. From your organization's Graphs page in [GraphOS Studio](https://studio.apollographql.com/?referrer=docs-content), open the PQL page for a graph by clicking its PQL button:

Reviewer 4: That's a good point about being able to provide a link. (The link you provide won't necessarily be ideal if users have multiple orgs but they don't in practice.)

I do think the other link will be used more (eg, it actually spells out what it is in words that you can search for on the page) but for the purpose of this tutorial, keeping something that you can provide a link to in the docs makes sense.

  <img
    className="screenshot"
    alt="The persisted query list button in the Studio graph list"
    src="../../img/pql-button.jpg"
  />

2. From the PQL page:

    - If you haven't created any PQLs yet, click **Create a Persisted Query List**.
    - If you already have at least one PQL, click **New List** in the upper right.

  The following dialog appears:

>   <img

Reviewer 4: (not a preview blocker imo)

    - This level provides no operation safelisting.
- **Audit mode**: Executing operations by providing a PQL-specified ID is still optional, but the router also logs any unregistered operations.
    - The level serves as a dry run and helps you identify operations you may still need to register before turning on safelisting.
- **Safelisting with operation strings allowed**: The router _rejects_ any incoming operations that aren't present in its PQL.
    - Before moving to this security level, ensure _all_ your client operations are present in your PQL.
- **Safelisting with IDs only**: The router rejects any incoming operations that aren't present in its PQL, _and_ clients _must_ execute an operation by providing its PQL-specified ID.
    - Before moving to this security level, ensure _all_ your clients execute operations by providing their PQL-specified ID.

See below for sample YAML configurations for each level.

#### Performance-only

To use persisted queries _only_ to reduce network bandwidth and latency (_not_ for safelisting), add the following minimal configuration:

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true

> enabled: true # APQ can be enabled

Reviewer 2: Artifact of `apq` enabled indeed. Removed all mention here: https://github.com/apollographql/docs/pull/496/commits/41b80906b9f14b50cfffa764d450110a3efcd7d9

```

> **Note:** You can use the performance-only security level with or without [automatic persisted queries](/router/configuration/in-memory-caching#caching-automatic-persisted-queries-apq) enabled.

This enables your clients to execute operations by providing their PQL-specified ID instead of the full operation string.
Your router also continues to accept full operation strings, even for operations that _don't_ appear in its PQL.

> The router works with both formats since the PQL-specified ID is the operation string's [SHA256 hash](https://en.wikipedia.org/wiki/SHA-2), and the router can derive it from the operation string if needed. 

Reviewer 4: Yeah, I'm not entirely sure what @martinbonnin meant there.

> The router works with both formats since the PQL-specified ID is the operation string's [SHA256 hash](https://en.wikipedia.org/wiki/SHA-2), and the router can derive it from the operation string if needed. 

Reviewer: My comment was related to the `Safelisting with operation strings allowed` mode. 

If the id is opaque, how does the router looks up if the operation is in the list? I was assuming there would be some kind of database with an `id` index. Or is the query string also indexed so that the lookup can happen based solely on the query string?

> The router works with both formats since the PQL-specified ID is the operation string's [SHA256 hash](https://en.wikipedia.org/wiki/SHA-2), and the router can derive it from the operation string if needed. 

Reviewer 4: Yeah, it just takes the mapping and inverts it (or well, makes a set of the values).

> The router works with both formats since the PQL-specified ID is the operation string's [SHA256 hash](https://en.wikipedia.org/wiki/SHA-2), and the router can derive it from the operation string if needed. 

Reviewer: Gotcha 👍 Apologies for the confusion.

#### Audit mode (dry run)

Turning on logging is crucial for gauging your client apps' readiness for safelisting. The logs identify which operations you need to either add to your PQL or stop your client apps from making. 

To enable logging for unregistered queries, enable the `log_unknown` property:

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
  log_unknown: true

> enabled: true # APQ can be enabled

Reviewer 4: ditto

Reviewer 2: https://github.com/apollographql/docs/pull/496/commits/41b80906b9f14b50cfffa764d450110a3efcd7d9

```
> **Note:** You can use audit mode with or without [automatic persisted queries](/router/configuration/in-memory-caching#caching-automatic-persisted-queries-apq) enabled.

Unregistered operations appear in your [router's logs](./logging).

#### Safelisting (operation strings allowed)

> ⚠️ Before applying this configuration, ensure your PQL contains _all_ of the GraphQL operations that _all active versions_ of your clients execute. If you enable safelisting _without_ ensuring this, your router will reject any unpublished client operations.

With the following configuration, your router allows _only_ GraphQL operations that are present in its PQL while rejecting all other operations:

```yaml title="router.yaml"
preview_persisted_queries:
  enabled: true
  log_unknown: true
  safelist:
    enabled: true
    require_id: false
apq:
  enabled: false # APQ must be disabled 
```

> > **Note:** To enable safelisting, you _must_ disable [automatic persisted queries](/router/configuration/in-memory-caching#caching-automatic-persisted-queries-apq).

Reviewer 4: This could get an explanation, like "that's because the point of APQs is that clients can register arbitrary operations on the fly, and the point of safelisting is that clients can't run arbitrary operations".

(One could imagine that if we allowed a per-request override of safelisting ("developer mode" or something) that we could allow those requests to use APQs... maybe. But it's not clear if we're going to do the former, or if it's helpful for them to be able to use APQs in this case.)

Reviewer 4: (not a preview blocker imo)

If you want to use this security level, you should always _first_ set up [safelisting with operation strings allowed](#safelisting-operation-strings-allowed). This level requires modifying _all_ of your clients to execute operations via PQL-specified ID instead of an operation string. While making those necessary changes, you can use the less restrictive safelisting level in your router.

### 3. Preregister trusted operations

Preregistering trusted operations to a PQL has two steps:
1. Generating persisted queries manifests (PQM) using client-specific tooling
2. Publishing PQMs to the PQL using the Rover CLI tool

Building both of these into your CI/CD pipeline incorporates new operations automatically whenever you release a new client app version.

#### 3.1 Generate persisted queries manifests

Once a PQL exists in GraphOS, client teams can start publishing operations to it. To do so, you need to generate JSON **manifests** of the trusted operations to publish. You generate a separate manifest for each of your client apps.

<ExpansionPanel title="See an example manifest file with two operations">

```json title="persisted-queries.json"
{  

>   "format": "apollo-persisted-queries",

Reviewer 4: ```suggestion
  "format": "apollo-persisted-query-manifest",
```

#### Generation methods

Apollo Client for Web, Kotlin, and iOS each provide a mechanism for generating a manifest file from your app source.

> If your client app uses another GraphQL client library, you can build your _own_ mechanism for generating operation manifests. See the expected [manifest format](#manifest-format).

See the instructions for your client library:

#### Apollo Client Web

1. In your app's project, install the `@apollo/generate-persisted-query-manifest` package as a dev dependency:

    ```bash
    npm install --save-dev @apollo/generate-persisted-query-manifest
    ```

    This package includes a CLI command to generate a manifest file from your application source.

2. Generate your first manifest with the following command:

    ```bash
    npx generate-persisted-query-manifest
    ```

    - **If the command succeeds,** your manifest is written to `persisted-query-manifest.json`.
    - **If the command fails** (or if your manifest doesn't include all of the operations you expect it to), you can configure the command's behavior using the options described in the [package README](https://www.npmjs.com/package/@apollo/generate-persisted-query-manifest).

> [See the full Apollo Client persisted queries guide for detailed instructions](/react/api/link/persisted-queries/).

Reviewer 4: As you know, this doesn't have information yet and we should do something else here for now.

Reviewer 2: https://github.com/apollographql/docs/pull/496/commits/b933b0238b9e5e0e75d3c17824425421ee3abacc

#### Apollo Kotlin

> Manifest generation requires Apollo Kotlin `3.8.2` or later.

To generate an operation manifest with Apollo Kotlin, you modify your project's [Gradle plugin configuration](/kotlin/advanced/plugin-configuration) to generate a manifest in addition to standard Kotlin source for your operations:

```kotlin
apollo {
  service("myapi") {
    packageName.set("com.example.myapi")
    operationManifestFormat.set("persistedQueryManifest") //highlight-line
  }
}
```
The manifest will be generated in `build/generated/manifest/apollo/myapi/persistedQueryManifest.json`

> [See the full Apollo Kotlin persisted queries guide for detailed instructions](/kotlin/advanced/persisted-queries/).

Reviewer 4: As you know, this doesn't have full information about this feature yet and we should do something for now until it does.

Reviewer 2: https://github.com/apollographql/docs/pull/496/commits/b933b0238b9e5e0e75d3c17824425421ee3abacc

</blockquote>

**The above command does the following:**

1. It publishes all operations in the provided manifest file to the PQL linked to the specified variant.
    - Publishing a manifest to a PQL is additive. Any _existing_ entries in the PQL remain.
    - If you publish an operation with the same `id` but different details from an existing entry in the PQL, the entire publish command fails with an error.

2. It updates any _other_ variants that the PQL is applied to so that routers associated with those variants can fetch their updated PQL.

### 4. Client updates

With your [manifest published](#3-preregister-trusted-operations) and the [router configured](#2-router-configuration), you can update your clients to use trusted operations IDs. Organizations can do this one client at a time as client teams [publish client-specific PQMs](#32-publish-manifests-to-the-pql) to the PQL.

> **Note:** This step provides performance benefits but isn't necessary for safelisting. You can continue to use full operation strings rather than operation IDs.

#### 4.1 Test operations

> Before updating your client apps, it's best to send some test operations.

Reviewer 4: I'm not sure how this is relevant to "client updates". It seems more like a thing you can do to test the "publish manifests" part.

Reviewer 2: https://github.com/apollographql/docs/pull/496/commits/683c11558d913d40ccf3dff1701c1cdd6d10ffb4

'{"extensions":{"persistedQuery":{"version":1,"sha256Hash":"dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f"}}}'
```

If your router's PQL includes an operation with an ID that matches the value of the provided `sha256Hash` property, it executes the corresponding operation and returns its result.

#### 4.2 Update client apps

Once you've tested requests using operation IDs, you can update your client apps to execute operations using their PQL-specified ID instead of sending full query strings.

To do this, clients can use the same protocol used for automatic persisted queries (APQ). 

Here's the JSON body of a request to execute an operation by its ID:

```json
{
  "variables": null,
  "extensions": {
    "persistedQuery": {
      "version": 1,
      "sha256Hash": "PQL_ID_HERE"
    }
  }
}
```

> If you're executing an operation that includes GraphQL variables, specify them with the `variables` property.

If your GraphQL client library provides a mechanism for executing operations via APQ, you can use that same mechanism with your router's PQL.

Here are links to documentation for APQ mechanisms in Apollo Client libraries:

> - [Apollo Client Web](/react/api/link/persisted-queries/)

Reviewer 4: ACW requires you to use this [new package](https://www.npmjs.com/package/@apollo/persisted-query-lists) at runtime in order to send the correct PQs (you use it along with the existing APQ client link).  In fact, you need to do this in order to turn on even ID-allowed safelisting, because certain aspects of the operation sent at runtime by ACW (notably, the order of fragment and operation definitions) may not match the registered operation unless you do so. This is different from iOS and Kotlin which have a much more deterministic approach to formatting operations.

Once you've confirmed all client apps use IDs, you can move to the most restrictive security level: [safelisting with IDs only](#safelisting-ids-only).
This security level _enforces_ the performance benefit of using operation IDs rather than full operation strings. If you're content with the safelisting aspect of persisted queries with only optional performance benefits, you don't need to enable it.

### Coordinate with client teams

Once you've followed the implementation steps for one client, you can coordinate across all your client teams:

1. Identify all the client apps that execute operations against your router, along with the GraphQL client libraries that those apps use.
    - Before you enable safelisting in your router, your client apps must start publishing their operations to your PQL.
2. Communicate to your client development teams that adopting persisted queries will require adding tooling to their CI/CD pipeline.

>     - Specifically, each client team will need to perform the following steps of this adoption path: [Generate client operation manifests](#31-generate-persisted-queries-manifests) and [publish manifests to the PQL](#32-publish-operations-to-the-pql)

Reviewer 4: For ACW specifically, they will also need to adopt one of the functions in [this new package](https://www.npmjs.com/package/@apollo/persisted-query-lists) (`generatePersistedQueryIdsAtRuntime` or `generatePersistedQueryIdsFromManifest`) at runtime. "just send freeform GraphQL without any runtime changes" is not an option for ACW if you want to turn on safelisting, for the normalization reason described above. (We may change safelisting to allow for slightly more "fuzzy" matching to avoid this requirement later.)

Reviewer 2: Step 2. is about integration PQs into their build pipeline, so Instead mentioning this above under "Update your chosen client to use only trusted operations"

Reviewer 4: The problem is that for ACW specifically, if you want to use *any* safelisting (not just "ID only") you will need to use the additional link. We just plain don't really support "safelisting, but you can send the operation as freeform GraphQL, but you're using GraphQL, with nothing special done to your client", due to fragment/operation ordering issues.

Honestly I think that this is probably an unacceptable overhead and we should improve the safelisting feature to ignore these operation ordering issues, so that "safelist but it's OK to send matching freeform without having to change runtime behavior" is actually usable with our most popular client.

3. Identify which team members will assist with adding tooling to their respective CI/CD pipelines.

```mermaid
graph LR;

  subgraph "CI/CD"
  webClient("Web client<br/>build");
  androidClient("Android client<br/>build");
  iosClient("iOS client<br/>build");
  end

  subgraph "GraphOS"
  pql[[Persisted<br/>Query List]];
  end

  webClient & androidClient & iosClient -->|Publishes<br/>operations|pql
```

Guide each client team to follow the implementation steps presented in the [incremental adoption path](#incremental-adoption-path).

## Manifest format

<blockquote>

⚠️ **This manifest format is subject to change during the preview period.**

You only need to read this section if you're building your own tooling to [generate persisted query manifests](#generating-manifests).

</blockquote>

A persisted query manifest has the following minimal structure:

```json title="persisted-queries-manifest.json"
{  

>   "format": "apollo-persisted-queries",

Reviewer 4: ```suggestion
  "format": "apollo-persisted-query-manifest",
```

  "version": 1,
  "operations": [  
    {
      "id": "dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f",
      "body": "query UniversalQuery { __typename }",
      "name": "UniversalQuery",
      "type": "query" 
    }
  ]
}
```

Manifest properties are documented below.

### Top-level properties

<table class="field-table">
  <thead>
    <tr>
      <th>Property</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr>
<td>

###### `format`

</td>
<td>

> This value is currently always `apollo-persisted-queries`.

Reviewer 4: ```suggestion
This value is currently always `apollo-persisted-query-manifest`.
```

</td>
</tr>

<tr>
<td>

###### `version`

</td>
<td>

This value is currently always `1`.

</td>
</tr>

<tr>
<td>

###### `operations`

</td>
<td>

An array of objects describing the individual GraphQL operations to publish.

For details, see [Per-operation properties](#per-operation-properties).

</td>
</tr>

</tbody>
</table>

### Per-operation properties

Each entry in a [manifest's](#manifest-format) `operations` array is a JSON object that describes a single GraphQL operation to publish:

```json
{
  "id": "dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f",
  "body": "query UniversalQuery { __typename }",
  "name": "UniversalQuery",
  "type": "query" 
}
```

Each operation object has the following properties:

<table class="field-table">
  <thead>
    <tr>
      <th>Property</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr>
<td>

###### `id`

</td>
<td>

The unique ID to use for the operation in your PQL.

> **This value must be unique among operations in the PQL.** If you publish an operation with the same `id` as an existing operation, the new operation _overwrites_ the existing PQL entry.

Reviewer 4: something like this maybe?

```suggestion
**This value must be unique among operations in the PQL.** It may match a previously-published operation with the same body. If you publish an operation with the same `id` as an existing operation but a different body, this is interpreted as an attempt to _overwrite_ the existing PQL entry with the new body, which will change the behavior of existing deployed clients. This will currently lead to an error when you publish the manifest.
```

Reviewer 2: Committing, but will rework.

Reviewer 4: Entire operation fails.

We do intend to allow you to say "no I really meant this" (our API supports it but rover doesn't), for moderately obscure use cases like "a client in some embedded hardware that you can't update is sending an ID, and you want that ID to start running a different operation instead", but we haven't made that part of the default UX for now.

<td>

The operation's name. Must match the name specified in `body`.

This value does _not_ need to be unique among operations in the PQL. Often, different clients execute slightly different operations with the same name, and those operations each require a separate entry in the PQL.

</td>
</tr>

<tr>
<td>

###### `type`

</td>
<td>

The type of GraphQL operation. Always one of the following values:

- `query`
- `mutation`
- `subscription`

</td>
</tr>

</tbody>
</table>

### Generating IDs

When generating operation IDs for a manifest, you should use a value that's unique to each operation, such as the query document's cryptographic hash. Apollo's [manifest generation tools](#13-generate-persisted-queries-manifests) use the base16 representation of the document's SHA256 hash, which is the same format used for [APQ](#differences-from-apq).

By generating identifiers based on query documents this way, you ensure that different operations always have different IDs. This prevents unexpected collisions in your PQL. It also allows the router to execute queries both by full query strings and PQL-specified IDs, since the router can derive the ID from the operation string if needed. 

> > **Never use an operation _names_ for PQL IDs.** Different clients (or even different versions of the _same_ client) might execute different operations with the same name, and _all_ of those distinct operations should be present in your PQL.

Reviewer 4: seems like this should be "never use an operation's name" or "never use operation names"?

Reviewer 2: https://github.com/apollographql/docs/pull/496/commits/889a3e7dc17abc70f754293603f98bd61851535c

Reviewer 2: https://github.com/apollographql/docs/pull/496/commits/889a3e7dc17abc70f754293603f98bd61851535c

### Ensuring consistent operation signatures

Whenever a client sends an operation string to a router that has safelisting enabled, the router checks for that operation string's presence in its persisted query list. **The incoming string must match a PQL entry's operation string _exactly_, including whitespace and field order.**

> The following queries are functionally identical, but the router will _reject_ the client operation because it doesn't match the PQL entry exactly:

Reviewer 4: Not exactly true tbh — the GraphQL server will return something like `{"data": {"books": [{"author": "A", "title": T"}]}}` for the first and like `{"data": {"books": [{"title": "T", "author": A"}]}}` for the latter. While these are treated equivalently by most applications, this order is actually specified by the system! Perhaps just use a comment or whitespace formatting here? Or use a fragment and show the difference between defining the fragment first or second, eg

```
fragment F on T {
  field
}
query Q {
  t {
    ...F
  }
}
```

it's different if you put the whole `fragment` after the whole `query` (which is one of the things that the `@apollo/persisted-query-list` functions keep consistent). But honestly maybe just a comment or whitespace change is fine here.

<CodeColumns>

```graphql title="PQL entry"
query GetBooks {
  books {
    author
    title
  }
}
```

```graphql title="Client operation"
query GetBooks {
  books {
    #highlight-start

>     title # Field order reversed!

Reviewer 4: in fact, the comment itself will also lead to rejection (which maybe you can mention explicitly, or maybe just leave the comment out?)

    </tr>
  </tbody>
</table>

If you _only_ want to improve request latency and bandwidth usage, APQ probably addresses your use case. If you _also_ want to secure your supergraph with operation safelisting, you should preregister persisted queries in a PQL.

## Security levels

The Apollo Router supports the following security levels, in increasing order of restrictiveness:

|Level|Description|
|--|--|
|**Performance-only**| Clients can optionally execute an operation on your router by providing the operation's PQL-specified ID.|
|**Auditing**| Executing operations by providing a PQL-specified ID is still optional, but the router logs any unregistered operations.|
|**Safelisting with query strings allowed**| The router _rejects_ any incoming operations that aren't present in its PQL.|
|**Safelisting with IDs only**| The router rejects any incoming operations that aren't present in its PQL, _and_ clients _must_ execute an operation by providing its PQL-specified ID.|

> _You can find more details, including configuration instructions, in the [implementation section](#2-router-configuration)._

> These levels of permissiveness allow you to [incrementally adopt](#incremental-adoption-path) persisted queries on a client-by-client basis. Specifically, the router should use auditing security level until you're confident that _all_ your clients' trusted operations have been registered in the PQL. Refer to the [incremental adoption section](#incremental-adoption-path) for a step-by-step guide.

Reviewer 2: ```suggestion
These levels of permissiveness allow you to [incrementally adopt](#incremental-adoption-path) persisted queries on a client-by-client basis. Specifically, the router should use the auditing security level until you're confident that _all_ your clients' trusted operations have been registered in the PQL. Refer to the [incremental adoption section](#incremental-adoption-path) for a step-by-step guide.
```

    src="../../img/pql-dialog-create.jpg"
    width="450"
  />

3. Provide a name and (optional) description for your PQL, then click **Create**.

    - At this point, your empty PQL has been created. The remaining dialog steps help with additional setup.

4. The second dialog step (**Link**) enables you to link your new PQL to one existing variant of your graph.

    - You can optionally **Skip** this step and link variants later (covered in the next step).
    
5. The third dialog step (**Publish**) displays your new PQL's unique ID and an example Rover CLI command for publishing operations to the PQL.

    - For now, we'll leave the PQL empty. We'll publish client operations to it in a later step.

6. The fourth and final dialog step (**Configure**) displays configuration options you apply to your router to begin using your PQL. We'll cover these in a later step.

7. Click **Finish** to close the dialog. Your newly created PQL appears in the table:

    <img
      className="screenshot"
      alt="A newly created PQL in the table on the Studio PQL page"
      src="../../img/pql-table.jpg"
      width="450"
    />

#### 1.2 Link the PQL to variants

> After you create a PQL, you can link it to one or more variants of your graph. Each router instance associated with a linked variant automatically fetches its PQL from GraphOS.

Reviewer 2: You can have multiple lists for testing but you can only link one PQL per variant.

Reviewer 2: Moot: To start, [all graphs will now have one automatically created PQL](https://apollograph.slack.com/archives/C04LHRSTLDD/p1690405777653289).

```bash
curl http://localhost:4000 -X POST --json \
'{"extensions":{"persistedQuery":{"version":1,"sha256Hash":"dc67510fb4289672bea757e862d6b00e83db5d3cbbcfb15260601b6f29bb2b8f"}}}'
```

If your router's PQL includes an operation with an ID that matches the value of the provided `sha256Hash` property, it executes the corresponding operation and returns its result.

#### 3.2 Update client apps

After your router is running with persisted queries enabled, you can update your client apps to execute operations using their PQL-specified ID instead of sending full query strings.

To do this, clients can use the same protocol used for automatic persisted queries (APQ). 

Here's the JSON body of a request to execute an operation by its ID:

```json
{
  "variables": null,
  "extensions": {
    "persistedQuery": {
      "version": 1,
      "sha256Hash": "PQL_ID_HERE"
    }
  }
}
```

> If you're executing an operation that includes GraphQL variables, specify them with the `variables` property.

If your GraphQL client library provides a mechanism for executing operations via APQ, you can use that same mechanism with your router's PQL.

> Here are links to documentation for APQ mechanisms in Apollo Client libraries:

Reviewer 2: Need further instructions here

1. Identify the first client you want to implement persisted queries with. It could be the client or team you're most comfortable with or the one most comfortable with GraphOS.
2. Follow all [implementation steps](#implementation-steps) for your chosen client:
    1. Begin by [creating a PQL in GraphOS](#11-pql-creation).
    1. When [linking the PQL](#12-link-the-pql-to-variants) to a variant, start with a testing or staging  environment.
    1. Use the [client-specific tooling](#generation-methods) to generate an [operation manifest](#13-generate-client-operation-manifests) as part of your client's CI/CD pipeline.
    1. [Publish the manifest](#14-publish-manifests-to-the-pql) to the PQL.
    1. When [configuring the router](#2-router-configuration), start with either [performance-only](#performance-only) or [auditing](#auditing-dry-run) security level.
    1. [Update your chosen client](#3-client-updates) to use only trusted operations, either by full query string or operation ID.

>     1. If you haven't already, [configure your router](#2-router-configuration) to use the [auditing](#auditing-dry-run) security level. Use the router logs to confirm that your the client is only sending trusted operations.

Reviewer 2: Not sure if there's more to monitoring logs than this. It seems like a studio observability section was in the works? (See https://apollographql.quip.com/59b0AlNDnjpQ/Safelisting-Architect-Narrative.)

Reviewer 2: Nope, only logs ATM.

    1. [Link the PQL](#12-link-the-pql-to-variants) to a production variant.

> 3. Continue to monitor your router logs: once you consistently see that unregistered operations are being logged and registered ones aren't, you've completed most of the setup for this client! 🎉

Reviewer 2: @BrynCooke , would you be able to provide examples of router logs for unregistered ops?

    - Before you enable safelisting in your router, your client apps must start publishing their operations to your PQL.
2. Communicate to your client development teams that adopting persisted queries will require adding tooling to their CI/CD pipeline.
    - Specifically, each client team will need to perform the following steps of this adoption path: [Generate client operation manifests](#13-generate-client-operation-manifests) and [publish manifests to the PQL](#14-publish-operations-to-the-pql)
3. Identify which team members will assist with adding tooling to their respective CI/CD pipelines.

```mermaid
graph LR;

  subgraph "CI/CD"
  webClient("Web client<br/>build");
  androidClient("Android client<br/>build");
  iosClient("iOS client<br/>build");
  end

  subgraph "GraphOS"
  pql[[Persisted<br/>Query List]];
  end

  webClient & androidClient & iosClient -->|Publishes<br/>operations|pql
```

Guide each client team to follow the implementation steps presented in the [incremental adoption path](#incremental-adoption-path).

## Safelisting as part of your security architecture

> Safelisting empowers you to reduce the surface of your supergraph by restricting the shape of incoming operations. It's one key component of GraphOS' security features and complements other security features such as [JWT authentication](/router/configuration/authn-jwt/) and [operation limits](/router/configuration/operation-limits).

Reviewer 2: Very open to feedback on the placement of this section. I originally had it closer to the beginning, but moved it closer to the end as it seems a nice transition to other security features. Happy to move it back.

Also: are these the correct features to mention here?

Reviewer 2: Removed. This doc is already too long, and this information would be better placed in a blog post.

        router <--> pql
    end

  subgraph Supergraph
  direction LR
  api1(["GraphQL API"]);
  api2(["GraphQL API"]);
  api3(["GraphQL API"]);

  end
  Apps -->|Queries| Router -->|Trusted<br/> queries| Supergraph
```

Your router can use its persisted query list (PQL) to both **protect your supergraph** and **speed up your clients' operations**:

- When you enable **safelisting**, your router _rejects_ any incoming operations not registered in its PQL.
- Client apps can execute an operation by providing its PQL-specified ID _instead of_ a full operation string.
    - Querying by ID can significantly reduce latency and bandwidth usage for very large operation strings.
    - Your router can _require_ clients to provide operations by ID and reject full operation strings (even operation strings in the PQL).

## Differences from automatic persisted queries

The Apollo Router also supports a related feature called [automatic persisted queries](/router/configuration/in-memory-caching#caching-automatic-persisted-queries-apq) (APQ). With APQ, clients can execute a GraphQL operation by sending the SHA256 hash of its operation string instead of the full string.

> APQ has a few limitations compared to preregistered persisted queries.

Reviewer 5: Really like this table!

      </td>
      <td>Platform team</td>
    </tr>
    <tr>
      <td> <a href="#3-preregister-trusted-operations">3. Preregister trusted operations</a> </td>
      <td>
      Generate and publish a persisted queries manifest (PQM) to the PQL from your client's CI/CD pipeline.
      </td>
      <td>App developers</td>
    </tr>
    <tr>
      <td><a href="#4-client-updates">4. Client updates (Optional) </a> </td>
      <td>
        Update clients to use operation IDs rather than full query strings.
        <br/>
        <br/>
        <em>This step provides performance benefits but isn't necessary for safelisting.</em>
      </td>
      <td>App developers</td>
    </tr>
  </tbody>
</table>

Continue reading for each step's details, or skip to the [incremental adoption section](#incremental-adoption-path) for the recommended incremental adoption strategy. (This section assumes you have a high-level understanding of each implementation step.)

### 1. PQL creation and linking

To use persisted queries, you first need a **persisted query list** (**PQL**) in GraphOS Studio.
Platform teams create an empty PQL in GraphOS Studio so that client teams can register trusted operations to it.

> Each PQL is associated with a single graph in GraphOS. A graph, however, can have several PQLs. For example, you may need multiple PQLs for one graph if want a separate PQL for each [contract variant](../delivery/contracts/). You can link a PQL to any variants of its graph. And although many variants may use the same PQL, each variant can only have one linked PQL at a time.

Reviewer 4: ```suggestion
Each PQL is associated with a single graph in GraphOS. A graph, however, can have several PQLs. For example, you may need multiple PQLs for one graph if you want a separate PQL for each [contract variant](../delivery/contracts/). You can link a PQL to any variants of its graph. And although many variants may use the same PQL, each variant can only have one linked PQL at a time.
```

### 4. Client updates

With your [manifest published](#3-preregister-trusted-operations) and the [router configured](#2-router-configuration), you can update your clients to use trusted operations IDs. Organizations can do this one client at a time as client teams [publish client-specific PQMs](#32-publish-manifests-to-the-pql) to the PQL.

> **Note:** This step provides performance benefits but isn't necessary for safelisting. You can continue to use full operation strings rather than operation IDs.

To execute operations using their PQL-specified ID instead of full operations strings, clients can use the same protocol used for automatic persisted queries (APQ). 

Here's the JSON body of a request to execute an operation by its ID:

```json
{
  "variables": null,
  "extensions": {
    "persistedQuery": {
      "version": 1,
      "sha256Hash": "PQL_ID_HERE"
    }
  }
}
```

> If executing an operation that includes GraphQL variables, specify them with the `variables` property.

Apollo's mobile clients let you use the same mechanism for executing persisted queries operations as APQs.
Refer to their persisted query documentation for implementation details.

> - [Apollo iOS](/ios/fetching/peristed-queries/)

Reviewer 2: Also hope to have these updated.

- [Apollo Kotlin](/kotlin/advanced/persisted-queries/)

The Apollo Client Web requires you to use an [additional package](https://www.npmjs.com/package/@apollo/persisted-query-lists) at runtime to send persisted queries. You use this package along with the [existing APQ client link](/react/api/link/persisted-queries). This package is required because certain aspects of the operation sent at runtime may not match the registered operations without it. Mobile clients have a more deterministic approach to formatting operations and, thus, don't need it.

> Refer to the Apollo Client Web's [persisted query documentation](/react/api/link/persisted-queries/) for implementation details.

Reviewer 2: Currently not updated with PQ instructions, but I hope to update this before launch.

Once your router's logs are clear of unexpected operations, you can configure your router to use the [safelisting security level](#router-security-levels). Then, [update your client apps](#4-client-updates) to use operation IDs rather than full query strings.

Once you've confirmed all client apps use IDs, you can move to the most restrictive security level: [safelisting with IDs only](#safelisting-ids-only).
This security level _enforces_ the performance benefit of using operation IDs rather than full operation strings. If you're content with the safelisting aspect of persisted queries with only optional performance benefits, you don't need to enable it.

### Coordinate with client teams

Once you've followed the implementation steps for one client, you can coordinate across all your client teams:

1. Identify all the client apps that execute operations against your router, and the GraphQL client libraries that those apps use.
    - Before you enable safelisting in your router, your client apps must start publishing their operations to your PQL.
2. Communicate to your client development teams that adopting persisted queries will require adding tooling to their CI/CD pipeline.

>     - Specifically, each client team will need to generate client operation manifests](#31-generate-persisted-queries-manifests) and [publish manifests to the PQL](#32-publish-operations-to-the-pql)

Reviewer 5: ```suggestion
    - Specifically, each client team will need to generate client operation [manifests](#31-generate-persisted-queries-manifests) and [publish manifests to the PQL](#32-publish-operations-to-the-pql)
```

Reviewer 5: Missing square bracket (please place where appropriate) causes the link to show up as text
